### PR TITLE
[None][feat] Support DSA MTP indexer top-k sharing

### PR DESCRIPTION
**Pull request by nvxuanyuc**
## Summary by CodeRabbit

* **New Features**
  * Added optional Top-K reuse for sparse attention during multi-token draft iterations, improving decoding efficiency.
  * Introduced a new configuration option to enable index sharing behavior for this workflow.
* **Bug Fixes**
  * Top-K results can now be preserved and reused across draft steps instead of being recomputed every time.
* **Tests**
  * Added coverage for the new Top-K reuse flow in multi-step decoding scenarios.

## Description

Implement GLM-5.2 IndexShare for the one-model MTP path: the indexer top-k indices are computed on the first draft step and reused for all following draft steps.

Based on #15574 for full GLM-5.2 IndexShare support - this PR covers MTP side.

A new `index_share_for_mtp_iteration` field is added to `DeepSeekSparseAttentionConfig`. It is resolved from the HF checkpoint config when left unset and threaded into the DSA indexer as `mtp_index_share`. On draft step 0 the per-request last-token top-k is stashed on the attention metadata; for draft steps > 0 the indexer skips the logits computation and the top-k selection, reusing the stashed indices via a copy instead.

## Test Coverage

**Unit test** — `tests/unittest/_torch/attention/sparse/dsa/test_dsa_indexer.py::test_indexer_decode_mtp_topk_reuse`: verifies that draft step 0 stores each request's last-token top-k and that later draft steps reuse it bit-exactly (no recompute). Covered for both the steady-state generation case and the first generation step right after prefill, at BS 1 and 2.

**E2E tests** — in `TestGLM52`:
  - `test_nvfp4_mtp_index_share` — GSM8K accuracy (GLM-5.2-NVFP4, CuteDSL MoE, MTP 3, TEP8). Measured 93.86.
  - `test_nvfp4_mtp_index_share_mtp_ar` — MTP acceptance-rate guard. Measured AR 62.75%.

**Accuracy A/B validation** (index-share in MTP is lossless).
  | Dataset | mtp3 index-share ON | mtp3 dense topk | mtp1 baseline |
  | --- | --- | --- | --- |
  | GSM8K (full) | 94.12 | 94.20 | 93.75 |
  | LongBench (20 samples) | 48.18 | 48.16 | 48.08 |

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by Superjomn — APPROVED** [review]
LGTM

**Review by yingguo-trt — APPROVED** [review]

**Review by xinhe-nv — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 1]
```diff
@@ -2826,6 +2851,25 @@ def sparse_attn_indexer(
...
+        # MTP Top-K stash: save each sequence's last-token Top-K for reuse
+        # by subsequent draft steps.
+        if (self.mtp_index_share
+                and getattr(metadata, "in_mtp_draft_loop", False)
+                and not reuse_topk):
+            rows = None
+            if num_generations > 0:
+                next_n = num_gen_tokens // num_generations
+                rows = topk_indices_buffer[num_ctx_tokens:num_ctx_tokens +
+                                           num_gen_tokens][next_n - 1::next_n]
```
Comment (xwang233) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: This stashes `topk_indices_buffer[...][next_n - 1::next_n]`, i.e. each gen request's last padded row. Every gen request is padded to `next_n = draft_len+1` rows for cuda-graph, so any row past `num_accepted` sits on the target's rejected-branch continuation. Under partial acceptance (the common case at k≥7), the reuse steps at i≥1 then attend a top-k that was selected by a rejected-branch query. After the kv-lens rewind those indices also point at stale rejected-branch KV, so every reuse step of a partially accepted round is corrupted.

vLLM handles this by compacting to each request's last accepted token before the reuse steps: after `set_skip_topk(True)` it calls `compact_topk_indices(token_indices_to_sample)`, where `token_indices_to_sample` are the last-accepted rows.

* [vllm/v1/spec_decode/llm_base_proposer.py#L598-L603](https://github.com/vllm-project/vllm/blob/6472131298e7b9ce81651cac90cc2dcc28963b55/vllm/v1/spec_decode/llm_base_proposer.py#L598-L603)
* [vllm/model_executor/models/deepseek_mtp.py#L196-L209](https://github.com/vllm-project/vllm/blob/6472131298e7b9ce81651cac90cc2dcc28963b55/vllm/model_executor/models/deepseek_mtp.py#L196-L209)

The fix here is to plumb `num_accepted_tokens` into the metadata and stash `base + (num_accepted-1).clamp(0, next_n-1)` per gen request. The context stash stays as is, and this is cuda-graph safe.

The same wrong-row pattern also shows up in the new `_enable_heuristic_topk` seed at line 2846. Lower severity there since it only warms up the next top-k rather than freezing it, but it wants the same `num_accepted-1` selection.

See an example patch: http://nv/jvo

Comment (nvxuanyuc) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Thanks! Have added the fix for partial acceptance indexing and factored into a helper that heuristic topk path can reuse.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py:1]: non-blocking NIT: these `getattr` seems unnecessary as the fields are required.

**Review by pengbowang-nv — APPROVED** [review]
Attention Part LGTM

File: tensorrt_llm/_torch/speculative/eagle3.py [hunk 2]
```diff
@@ -839,8 +839,17 @@ def _forward_linear_draft_loop(self, inputs, attn_metadata, spec_metadata,
             attn_metadata.seq_lens_cuda, dim=0, dtype=torch.long) - 1
         position_ids = inputs["position_ids"]
 
+        reuse_mtp_topk = (self.is_mtp_eagle
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/speculative/eagle3.py:842]: Should we also add the mtp_index_share check here too？

**Review by zhaoyangwang-nvidia — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py
Comment (nvxuanyuc) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py:1]: Thanks! Have switched the following fields to direct attribute access, including `indexer_skip_topk`, `shared_topk_indices`, and `in_mtp_draft_loop`.
Two `getattr` are still kept:
- `mtp_num_accepted`: the indexer unit test drives this path with a lightweight mock metadata that doesn't populate this field.
- `mtp_index_share`: DeepSeekV4 shares the indexer with its own sparse-config class that doesn't define this field.

File: tensorrt_llm/_torch/speculative/eagle3.py (hunk 2 above)
Comment (nvxuanyuc) [reply tensorrt_llm/_torch/speculative/eagle3.py:842]: All the other index-share logic is already safely gated by `mtp_index_share` in `dsa.py` (e.g. `reuse_topk`); the only missing guard was at the `shared_topk_indices` stash in `mla.py` which I have added now. I chose not to plumb `mtp_index_share` up to eagle3's `reuse_mtp_topk` here, since that would need additional exposure of the indexer on the metadata.

File: tensorrt_llm/llmapi/llm_args.py [hunk 3]
```diff
@@ -918,6 +918,11 @@ class DeepSeekSparseAttentionConfig(SeqLenAwareSparseAttentionConfig):
         "`fp4` requires Blackwell+ (SM>=100) at runtime and "
         "index_head_dim=128.",
     )
+    index_share_for_mtp_iteration: Optional[bool] = Field(
+        default=None,
+        description=
+        "Reuse the indexer Top-K across MTP draft steps instead of recomputing "
+        "it each step. Defaults to the model's HF config value.")
```
Comment (juney-nvidia) [inline tensorrt_llm/llmapi/llm_args.py:925]: For new flag, pls mark this explicitly as "prototype" so there will be less commitment about the API compatibility and also less maintenance ovheread.

Comment (juney-nvidia) [reply tensorrt_llm/llmapi/llm_args.py:925]: Let's do the adjustment in the subsequent PR to unblock firstly.

**Review by juney-nvidia — APPROVED** [review]
Approved from API/telemetry/OSS compliance perspective, and for newly added flag, pls explicitly declare as "prototype" in the subsequent PR.

File: tensorrt_llm/llmapi/llm_args.py (hunk 3 above)
Comment (Tabrizian) [reply tensorrt_llm/llmapi/llm_args.py:925]: @nvxuanyuc Please make this change in follow up PR. Resolving to unblock merging.

Comment (nvxuanyuc) [reply tensorrt_llm/llmapi/llm_args.py:925]: Thanks for the reminder! Fix in #17149.
